crates/bevy_animation/src/lib.rs

```diff
@@ -192,100 +243,161 @@ pub fn animation_player(
-                        if curve.keyframe_timestamps.len() == 1 {
-                            match &curve.keyframes {
-                                Keyframes::Rotation(keyframes) => transform.rotation = keyframes[0],
-                                Keyframes::Translation(keyframes) => {
-                                    transform.translation = keyframes[0];
-                                }
-                                Keyframes::Scale(keyframes) => transform.scale = keyframes[0],
-                            }
-                            continue;
-                        }
+        _update_transitions(&mut player, &time);
+        _run_animation_player(
+            entity,
+            player,
+            &time,
+            &animations,
+            &names,
+            &mut transforms,
+            &children,
+        );
+    }
+}
 
-                        // Find the current keyframe
-                        // PERF: finding the current keyframe can be optimised
-                        let step_start = match curve
-                            .keyframe_timestamps
-                            .binary_search_by(|probe| probe.partial_cmp(&elapsed).unwrap())
-                        {
-                            Ok(n) if n >= curve.keyframe_timestamps.len() - 1 => continue, // this curve is finished
-                            Ok(i) => i,
-                            Err(0) => continue, // this curve isn't started yet
-                            Err(n) if n > curve.keyframe_timestamps.len() - 1 => continue, // this curve is finished
-                            Err(i) => i - 1,
-                        };
-                        let ts_start = curve.keyframe_timestamps[step_start];
-                        let ts_end = curve.keyframe_timestamps[step_start + 1];
-                        let lerp = (elapsed - ts_start) / (ts_end - ts_start);
-
-                        // Apply the keyframe
+fn _run_animation_player(
+    entity: Entity,
+    mut player: Mut<AnimationPlayer>,
+    time: &Time,
+    animations: &Assets<AnimationClip>,
+    names: &Query<&Name>,
+    transforms: &mut Query<&mut Transform>,
+    children: &Query<&Children>,
+) {
+    let paused = player.paused;
+    // Continue if paused unless the `AnimationPlayer` was changed
+    // This allow the animation to still be updated if the player.elapsed field was manually updated in pause
+    if paused && !player.is_changed() {
+        return;
+    }
+
+    // Apply the main animation
+    _apply_animation(
+        1.0,
+        &mut player.animation,
+        paused,
+        entity,
+        time,
+        animations,
+        names,
+        transforms,
+        children,
+    );
+
+    // Apply any potential fade-out transitions from previous animations
+    for AnimationTransition {
+        current_weight,
+        animation,
+        ..
+    } in &mut player.transitions
+    {
+        _apply_animation(
```

Comment:
This redoes the hierarchy query for the same bones repeatedly, which can be the most expensive part of applying animations as it traverses the hierarchy and does name comparisons at every level. Can we move the animation sample to the inner part of the loop instead of into the outside?

Reply:
Hm good point. It's not very straightforward though because different animations may attach to different bones, so I cannot just make the bones the outer loop and the animations the inner loop. I could first calculate a union set of bones over all animations and then iterate over that in the outer loop? That would add some overhead for the calculation of the union set though. What do you think?

Reply:
We could also store an `Entity` id instead of an `EntityPath` in `AnimationClip`. `Entity` id shouldn't change after things are instantiated. Or would that break something?

Reply:
If an entity's `Name` or parent/children change, the path no longer is valid. #6785 introduces a per-path cache of entities to reduce the work needed for searching for the target entity. We could do something like that to reduce the load from the searches. That probably could work until the full animation composition RFC is merged and implemented, which has a more holistic solution to this.

Reply:
That works for me. We can merge this PR and then when #6785 gets merged, we'll automatically have the cache.